riscv/src/asm.rs

```diff
@@ -98,38 +110,18 @@ instruction!(
 /// are ordinarily not ordered with respect to loads and stores in the instruction stream.
 /// Executing an `SFENCE.VMA` instruction guarantees that any stores in the instruction stream prior to the
 /// `SFENCE.VMA` are ordered before all implicit references subsequent to the `SFENCE.VMA`.
-#[inline]
-#[allow(unused_variables)]
-pub unsafe fn sfence_vma(asid: usize, addr: usize) {
-    match () {
-        #[cfg(riscv)]
-        () => core::arch::asm!("sfence.vma {0}, {1}", in(reg) addr, in(reg) asid),
-
-        #[cfg(not(riscv))]
-        () => unimplemented!(),
-    }
-}
-
-/// `ECALL` instruction wrapper
-///
-/// Generates an exception for a service request to the execution environment.
-/// When executed in U-mode, S-mode, or M-mode, it generates an environment-call-from-U-mode
-/// exception, environment-call-from-S-mode exception, or environment-call-from-M-mode exception,
-/// respectively, and performs no other operation.
-///
-/// # Note
-///
-/// The ECALL instruction will **NOT** save and restore the stack pointer, as it triggers an exception.
-/// The stack pointer must be saved and restored accordingly by the exception handler.
-#[inline]
-pub unsafe fn ecall() {
-    match () {
-        #[cfg(riscv)]
-        () => core::arch::asm!("ecall", options(nostack)),
-
-        #[cfg(not(riscv))]
-        () => unimplemented!(),
-    }
+#[inline(always)]
+#[cfg_attr(
+    not(any(target_arch = "riscv32", target_arch = "riscv64")),
+    allow(unused_variables)
+)]
+pub fn sfence_vma(asid: usize, addr: usize) {
+    #[cfg(any(target_arch = "riscv32", target_arch = "riscv64"))]
+    unsafe {
+        core::arch::asm!("sfence.vma {0}, {1}", in(reg) addr, in(reg) asid, options(nostack));
+    };
+    #[cfg(not(any(target_arch = "riscv32", target_arch = "riscv64")))]
+    unimplemented!();
```

Comment:
This can be expressed using the `instruction` macro.

I have an example of how to do it here: https://github.com/rust-embedded/riscv/compare/asm...rmsyn:riscv:asm

Reply:
It didn't work on my machine, looks like it matches the first macro branch and assumes all the tokens correspond to the `options` stream.